extensions/amp-access-scroll/0.1/read-depth-tracker.js

```diff
@@ -0,0 +1,103 @@
+/**
+ * Copyright 2018 The AMP HTML Authors. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS-IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import {PositionObserverFidelity}
+  from '../../../src/service/position-observer/position-observer-worker';
+import {RelativePositions} from '../../../src/layout-rect';
+import {Services} from '../../../src/services';
+import {debounce} from '../../../src/utils/rate-limit';
+import {getServiceForDoc} from '../../../src/service';
+import {installPositionObserverServiceForDoc}
+  from '../../../src/service/position-observer/position-observer-impl';
+
+export class ReadDepthTracker {
+  /**
+   * Sets up tracking of paragraphs in document.
+   * @param {!../../../src/service/ampdoc-impl.AmpDoc} ampdoc
+   * @param {!../../amp-access/0.1/amp-access-source.AccessSource} accessSource
+   * @param {string} connectHostname
+   */
+  constructor(ampdoc, accessSource, connectHostname) {
+    /** @private {!../../../src/service/ampdoc-impl.AmpDoc} */
+    this.ampdoc_ = ampdoc;
+
+    /** @private {string?} */
+    this.lastIntersectingSnippet_ = null;
+
+    /** @private {!../../amp-access/0.1/amp-access-source.AccessSource} */
+    this.accessSource_ = accessSource;
+
+    /** @private {string} */
+    this.connectHostname_ = connectHostname;
+
+    /** @private {function(string)} */
+    this.debouncedUpdateLastRead_ = debounce(
+        ampdoc.win,
+        this.updateLastRead.bind(this),
+        1000
+    );
+
+    installPositionObserverServiceForDoc(ampdoc);
```

Comment:
yeah, the service is not in core but I don't think position-observer should be used here here exact location does not matter, just monitoring paragraphs with `IntersectionObserver` Dima mentioned is better.

Adding @zhouyx to this PR since this is analytics related and should follow 3P analytics hooks we have already.

Reply:
Unfortunately IntersectionObserver doesn't work well because it doesn't tell if the element is above or below viewport. The analytics team had a meeting regarding this. https://github.com/ampproject/amphtml/issues/20279#issuecomment-454498628

In short, we believe the use case here is very specific. And current analytics features couldn't fulfill the requirements here. Happy to discuss more if there're other concerns. Thanks.

Reply:
Sounds like we're okay with leaving this given the analytics conversation, but let me know if there are further issues here.

Reply:
Hello @helin24 I know we agreed to not use `InteresectionObserver` here. But I am concerning about performance if we are going to use `PositionObserver`. Even with low fidelity, it will try to calculate all paragraphs' `#getBoundingClientRect()` every 5 animation frames. 

Can we only calculate position at `#Viewport.onChange()`. There're `#Viewport.getClientRectAsync(el)` and `#Viewport.getLayoutRect(el)` to calculate non-AMP element's position apart from positionObserver.

Reply:
@zhouyx thanks for this suggestion - I updated to use `Viewport` instead of `PositionObserver`